# Lecture Link - https://www.youtube.com/playlist?list=PLl1irxoYh2wyLwJutUZx5Q_QEEDZoXBnz

# Code - https://github.com/blackdew/ml-tensorflow

# 지도학습
## 순서
### 1. 과거의 데이터 준비
### 2. 모델의 구조 만들기
### 3. 데이터로 모델을 학습(fit) 시키기
### 4. 모델 이용
#### Fit -> 모델을 데이터에 맞게 피팅하는 것(학습)

##### 원인이 되는 변수 -> 독립 변수
##### 결과가 되는 변수 -> 종속 변수

In [ ]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import tensorflow as tf

# co_accuracy = pd.read_csv('C:\\Users\\hhshi\\OneDrive\\바탕 화면\\CO오염도_최종.csv',encoding = 'utf-8')
lemonade = pd.read_csv('C:\\신현호\\2022년_머러공부\\youtube\\dataset\\lemonade.csv',encoding='utf-8')
boston = pd.read_csv('C:\\신현호\\2022년_머러공부\\youtube\\dataset\\boston.csv')
iris = pd.read_csv('C:\\신현호\\2022년_머러공부\\youtube\\dataset\\iris.csv')

In [ ]:
print(lemonade.columns)
print(boston.columns)
print(iris.columns)

Index(['온도', '판매량'], dtype='object')
Index(['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax',
       'ptratio', 'b', 'lstat', 'medv'],
      dtype='object')
Index(['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭', '품종'], dtype='object')


# Lemonade sale prediction

In [ ]:
# 독립, 종속 변수 나누기 (컬럼 별로 값 가져오기)
# lemonade
ind_var = lemonade[['온도']]
dep_var = lemonade[['판매량']]
print(ind_var)
print(dep_var)
type(ind_var)

   온도
0  20
1  21
2  22
3  23
4  24
5  25
   판매량
0   40
1   42
2   44
3   46
4   48
5   50


pandas.core.frame.DataFrame

In [ ]:
# Making structure of model
X = tf.keras.layers.Input(shape=[1]) # 숫자가 1인 이유 -> 독립변수가 한 개이기 때문이다.
Y = tf.keras.layers.Dense(1)(X) # 숫자가 1인 이유 -> 마찬가지로 종속 변수가 한 개이기 때문이다.
model = tf.keras.models.Model(X,Y)
model.compile(loss='mse')

# Loss
## (독립 변수를 모델에 넣어서 나온 예측값들- 실제 결과)^2 들의 평균 값 (mse 사용 시)
## 따라서, Loss가 0에 가까워질수록 학습이 잘 된 모델

In [ ]:
model.fit(ind_var, dep_var, epochs=1000,verbose=0)  # epochs -> 모델을 몇 번 학습할 지
# verbose를 0으로 설정하면 일일히 출력값을 출력하는 일이 없다.

In [ ]:
model.predict(ind_var)

array([[39.989586],
       [41.98905 ],
       [43.988518],
       [45.98798 ],
       [47.987446],
       [49.98691 ]], dtype=float32)

In [ ]:
dep_var

판매량
0   40
1   42
2   44
3   46
4   48
5   50

In [ ]:
model.predict([[15]]) # 15도일 때의 판매량 예측

array([[29.992268]], dtype=float32)

# Boston housing price prediction

In [ ]:
boston.columns

Index(['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax',
       'ptratio', 'b', 'lstat', 'medv'],
      dtype='object')

In [ ]:
boston = boston.dropna(axis=1)
boston

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     ptratio       b  lstat  medv  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       17.8  392.83   4.03  34.7  
3       18.7  394.63   2.94  33.4  
4       18.7  396.90   5.33  36.2  
..       ...     ...    ...   ...  
501     21.0  391.99   9.67  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9  

[506 rows x 14 columns]

In [ ]:
ind_var_boston = boston[['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax',
       'ptratio', 'b', 'lstat']]
dep_var_boston = boston[['medv']]
print(ind_var_boston.shape, dep_var_boston.shape)

(506, 13) (506, 1)


In [ ]:
# 모델 구조
X = tf.keras.layers.Input(shape=[13])
Y = tf.keras.layers.Dense(1)(X) 
model_bos = tf.keras.models.Model(X,Y)
model_bos.compile(loss='mse')

# 13개의 입력으로 1개의 출력을 만들어내는 구조
# 13개의 입력층, 1개의 출력층
# y = w1x1 + w2x2 + .... + w13x13+b

# 만약 12개의 입력층과 2개의 출력층이라면?
# y1 = w1x1+ .... + w12x12 + b
# y2 = w1x1 + ... + w12x12 + b 

# 첫번 째 수식에서 w 12개, b 1개
# 두번째 수식에서 w 12개, b 1개 총 26개의 숫자 찾아야함

In [ ]:
# 모델 fit 하기
model_bos.fit(ind_var_boston, dep_var_boston,epochs=1000)

Epoch 1/1000
16/16 [==============================] - 0s 600us/step - loss: 23.0355
Epoch 2/1000
16/16 [==============================] - 0s 666us/step - loss: 23.2641
Epoch 3/1000
16/16 [==============================] - 0s 600us/step - loss: 22.8060
Epoch 4/1000
16/16 [==============================] - 0s 467us/step - loss: 23.2717
Epoch 5/1000
16/16 [==============================] - 0s 533us/step - loss: 23.0567
Epoch 6/1000
16/16 [==============================] - 0s 800us/step - loss: 23.1733
Epoch 7/1000
16/16 [==============================] - 0s 734us/step - loss: 22.5783
Epoch 8/1000
16/16 [==============================] - 0s 533us/step - loss: 23.4687
Epoch 9/1000
16/16 [==============================] - 0s 533us/step - loss: 23.2211
Epoch 10/1000
16/16 [==============================] - 0s 667us/step - loss: 22.6950
Epoch 11/1000
16/16 [==============================] - 0s 667us/step - loss: 23.5244
Epoch 12/1000
16/16 [==============================] - 0s 533us/step - los

In [ ]:
model_bos.predict(ind_var_boston[0:5])

array([[29.205477],
       [24.21123 ],
       [30.376736],
       [28.764305],
       [28.282923]], dtype=float32)

In [ ]:
boston[['medv']][0:5]

medv
0  24.0
1  21.6
2  34.7
3  33.4
4  36.2

In [ ]:
# 모델 수식 확인
model_bos.get_weights()  # 가중치 가져오기

[array([[-9.7691655e-02],
        [ 4.6617780e-02],
        [-3.6868413e-03],
        [ 2.6346443e+00],
        [-8.5219727e+00],
        [ 4.7443485e+00],
        [-6.1575826e-03],
        [-1.2278856e+00],
        [ 2.3755276e-01],
        [-1.2053286e-02],
        [-6.8141276e-01],
        [ 1.1240847e-02],
        [-4.8472962e-01]], dtype=float32),
 array([18.898045], dtype=float32)]

In [ ]:
# arr = model_bos.get_weights()
# weight = arr[0]
# weight = weight.tolist()
# weight = sum(weight,[])

# res = 0

# for i in weight:
# #     res+=i
#     print(i)
# print(weight)

# Iris classification
## 품종을 종속 변수로 지정하는데, 위에서와는 달리 수치형 변수가 아닌 범주형 변수이기 때문에 회귀 알고리즘이 아닌, 분류 알고리즘을 사용해야한다.

In [ ]:
iris
# 꽃잎 -> petal, 꽃받침 -> sepal

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭         품종
0     5.1  3.5    1.4   0.2     setosa
1     4.9  3.0    1.4   0.2     setosa
2     4.7  3.2    1.3   0.2     setosa
3     4.6  3.1    1.5   0.2     setosa
4     5.0  3.6    1.4   0.2     setosa
..    ...  ...    ...   ...        ...
145   6.7  3.0    5.2   2.3  virginica
146   6.3  2.5    5.0   1.9  virginica
147   6.5  3.0    5.2   2.0  virginica
148   6.2  3.4    5.4   2.3  virginica
149   5.9  3.0    5.1   1.8  virginica

[150 rows x 5 columns]

# Onehot-encoding
## 딥러닝 모델을 사용하기 위해서는 모든 범주형 변수를 Onehot-encoding 해줘야함.
### -> pd.get_dummies 사용

In [ ]:
# Onehot-encoding
# get_dummies를 사용하면 데이터 내의 범주형 변수들만 골라서 원핫인코딩 된 결과 만들어줌
iris = pd.get_dummies(iris)

In [ ]:
iris

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭  품종_setosa  품종_versicolor  품종_virginica
0     5.1  3.5    1.4   0.2          1              0             0
1     4.9  3.0    1.4   0.2          1              0             0
2     4.7  3.2    1.3   0.2          1              0             0
3     4.6  3.1    1.5   0.2          1              0             0
4     5.0  3.6    1.4   0.2          1              0             0
..    ...  ...    ...   ...        ...            ...           ...
145   6.7  3.0    5.2   2.3          0              0             1
146   6.3  2.5    5.0   1.9          0              0             1
147   6.5  3.0    5.2   2.0          0              0             1
148   6.2  3.4    5.4   2.3          0              0             1
149   5.9  3.0    5.1   1.8          0              0             1

[150 rows x 7 columns]

# setosa, versicolor, virginica 세 가지 품종에 대한 수식을 만들어야함
# y1 = w1x1+w2x2+.... + b -> setosa
# y2 = w1x1 + ... + b -> versicolor
# y3 = w1x1 + ... + b -> virginica

## 따라서, 컴퓨터는 세 가지 품종에 대한 가중치를 모두 찾아야함

# Softmax
## 비율 예측을 위해 activation function 중 하나인 Softmax 사용
## activation function 역할: 퍼셉트론의 출력이 어떤 형태로 나가야하는지를 조절해줌

## 회귀에 사용하게 되는 loss는 mse, 분류에 사용하는 loss는 categorical_crossentropy이다.

In [ ]:
iris.columns

Index(['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭', '품종_setosa', '품종_versicolor',
       '품종_virginica'],
      dtype='object')

In [ ]:
ind_var_iris = iris[['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭']]
dep_var_iris = iris[['품종_setosa', '품종_versicolor',
       '품종_virginica']]

In [ ]:
X = tf.keras.layers.Input(shape=[4])
Y = tf.keras.layers.Dense(3,activation='softmax')(X)
model_iris = tf.keras.models.Model(X,Y)
model_iris.compile(loss = 'categorical_crossentropy',metrics='accuracy')

In [ ]:
model_iris.fit(ind_var_iris, dep_var_iris,epochs=100)

Epoch 1/100
5/5 [==============================] - 0s 2ms/step - loss: 0.8456 - accuracy: 0.4867
Epoch 2/100
5/5 [==============================] - 0s 1ms/step - loss: 0.8401 - accuracy: 0.4800
Epoch 3/100
5/5 [==============================] - 0s 1ms/step - loss: 0.8343 - accuracy: 0.4867
Epoch 4/100
5/5 [==============================] - 0s 2ms/step - loss: 0.8302 - accuracy: 0.5067
Epoch 5/100
5/5 [==============================] - 0s 1ms/step - loss: 0.8247 - accuracy: 0.5133
Epoch 6/100
5/5 [==============================] - 0s 1000us/step - loss: 0.8192 - accuracy: 0.4933
Epoch 7/100
5/5 [==============================] - 0s 1ms/step - loss: 0.8140 - accuracy: 0.5000
Epoch 8/100
5/5 [==============================] - 0s 999us/step - loss: 0.8087 - accuracy: 0.5067
Epoch 9/100
5/5 [==============================] - 0s 2ms/step - loss: 0.8045 - accuracy: 0.5133
Epoch 10/100
5/5 [==============================] - 0s 1ms/step - loss: 0.7984 - accuracy: 0.5067
Epoch 11/100
5/5 [======

In [ ]:
# 처음에 있는 5개의 데이터를 가지고 예측
model_iris.predict(ind_var_iris[0:5])
# 퍼센티지가 가장 높은 쪽으로 꽃 종류 분류

array([[0.8712439 , 0.08247293, 0.04628321],
       [0.7988932 , 0.12082543, 0.08028137],
       [0.84736276, 0.09864279, 0.05399444],
       [0.8014828 , 0.1275649 , 0.07095232],
       [0.88317704, 0.07734422, 0.03947882]], dtype=float32)

In [ ]:
print(dep_var_iris[0:5])

   품종_setosa  품종_versicolor  품종_virginica
0          1              0             0
1          1              0             0
2          1              0             0
3          1              0             0
4          1              0             0


In [ ]:
# 끝에 있는 5개 예측
model_iris.predict(ind_var_iris[-5:])

array([[0.0316449 , 0.43595567, 0.5323994 ],
       [0.0253917 , 0.40595078, 0.5686575 ],
       [0.03473356, 0.44396582, 0.5213007 ],
       [0.04200952, 0.5301942 , 0.42779627],
       [0.04301732, 0.49600676, 0.46097597]], dtype=float32)

In [ ]:
print(dep_var_iris[-5:])

     품종_setosa  품종_versicolor  품종_virginica
145          0              0             1
146          0              0             1
147          0              0             1
148          0              0             1
149          0              0             1


In [ ]:
# 가중치 출력
model_iris.get_weights()

[array([[-0.33676526, -0.45038068, -0.0144235 ],
        [ 0.6406141 , -0.2510822 , -0.763229  ],
        [-0.05308026,  1.0081924 ,  1.0501297 ],
        [-0.22489521,  0.10058852, -0.05913624]], dtype=float32),
 array([ 0.6190525 ,  0.41109788, -0.62422806], dtype=float32)]

# 위의 결과를 참고해보면
## setosa 의 공식은
## y1 = -0.33676526x1 + 0.6406141x2 + (-0.05308026x3) + (-0.22489521)x4 +  0.6190525

### 이런 식으로 위의 값들을 참고하여 나머지 두 꽃의 종류에 대해서도 y 식을 만들 수 있다.

# Hidden layer
## 보스턴 집값 모델 구조 예시 들기
## 은닉 계층 사용 시, 좀 더 나은 모델을 만들 수 있다.

In [ ]:
# X = tf.keras.layers.Input(shape=[13])
# H = tf.keras.layers.Dense(5, activation = 'swish')(X) # 히든 레이어 생성 5개의 노드 가짐
# Y = tf.keras.layers.Dense(1)(H) # 위에선 X로 했지만 이 경우엔 Hidden layer가 중간에 있으므로 H로 실행
# model = tf.keras.models.Model(X,Y)
# model.compile(loss='mse')

In [ ]:
# 모델 구조
X = tf.keras.layers.Input(shape=[13])
H = tf.keras.layers.Dense(10, activation = 'swish')(X) # 히든 레이어 생성 10개의 노드 가짐
Y = tf.keras.layers.Dense(1)(H) 
model_h = tf.keras.models.Model(X,Y)
model_h.compile(loss='mse')

# 모델 fit 하기
model_h.fit(ind_var_boston, dep_var_boston,epochs=1000, verbose=0)
model_h.fit(ind_var_boston, dep_var_boston,epochs=1000)

In [ ]:
model_h.summary()

Model: "model_10"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_11 (InputLayer)        [(None, 13)]              0         
_________________________________________________________________
dense_11 (Dense)             (None, 10)                140       
_________________________________________________________________
dense_12 (Dense)             (None, 1)                 11        
Total params: 151
Trainable params: 151
Non-trainable params: 0
_________________________________________________________________


In [ ]:
# 예측값과 실제값 비교
print(model_h.predict(ind_var_boston[:5]))
print(dep_var_boston[:5])

[[29.33384 ]
 [24.008694]
 [30.60481 ]
 [29.43706 ]
 [28.945059]]
   medv
0  24.0
1  21.6
2  34.7
3  33.4
4  36.2


# Iris classification using hidden layers

In [ ]:
X = tf.keras.layers.Input(shape=[4])
H1 = tf.keras.layers.Dense(8, activation = 'swish')(X)
H2 = tf.keras.layers.Dense(8, activation = 'swish')(H1)
H3 = tf.keras.layers.Dense(8, activation = 'swish')(H2)
Y = tf.keras.layers.Dense(3,activation='softmax')(H3)
model_iris_h = tf.keras.models.Model(X,Y)
model_iris_h.compile(loss = 'categorical_crossentropy',metrics='accuracy')

model_iris_h.fit(ind_var_iris, dep_var_iris,epochs=100)

Epoch 1/100
5/5 [==============================] - 0s 2ms/step - loss: 1.0842 - accuracy: 0.5667
Epoch 2/100
5/5 [==============================] - 0s 2ms/step - loss: 1.0386 - accuracy: 0.6600
Epoch 3/100
5/5 [==============================] - 0s 2ms/step - loss: 1.0102 - accuracy: 0.6600
Epoch 4/100
5/5 [==============================] - 0s 2ms/step - loss: 0.9837 - accuracy: 0.6600
Epoch 5/100
5/5 [==============================] - 0s 2ms/step - loss: 0.9579 - accuracy: 0.6733
Epoch 6/100
5/5 [==============================] - 0s 2ms/step - loss: 0.9311 - accuracy: 0.6867
Epoch 7/100
5/5 [==============================] - 0s 3ms/step - loss: 0.9040 - accuracy: 0.6867
Epoch 8/100
5/5 [==============================] - 0s 2ms/step - loss: 0.8780 - accuracy: 0.7533
Epoch 9/100
5/5 [==============================] - 0s 2ms/step - loss: 0.8505 - accuracy: 0.8200
Epoch 10/100
5/5 [==============================] - 0s 2ms/step - loss: 0.8252 - accuracy: 0.7600
Epoch 11/100
5/5 [===========

In [ ]:
print(model_iris_h.predict(ind_var_iris[:5]))
print(dep_var_iris[:5])

[[9.9986959e-01 1.3036665e-04 3.2253006e-10]
 [9.9878973e-01 1.2103285e-03 6.7669834e-09]
 [9.9964035e-01 3.5966377e-04 2.0987754e-09]
 [9.9920470e-01 7.9535437e-04 7.4284316e-09]
 [9.9991715e-01 8.2834711e-05 2.2367309e-10]]
   품종_setosa  품종_versicolor  품종_virginica
0          1              0             0
1          1              0             0
2          1              0             0
3          1              0             0
4          1              0             0


# 데이터 타입 조정
## 변수를 범주형 데이터로 바꾸기

In [ ]:
path = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/iris2.csv'
new_iris = pd.read_csv(path)
new_iris
# 품종 수치화 데이터를 범주형 데이터로 바꿔줘야함
# 이 경우엔 원래 쓰던 pd.get_dummies를 처음부터 쓸 수 없다.

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭  품종
0     5.1  3.5    1.4   0.2   0
1     4.9  3.0    1.4   0.2   0
2     4.7  3.2    1.3   0.2   0
3     4.6  3.1    1.5   0.2   0
4     5.0  3.6    1.4   0.2   0
..    ...  ...    ...   ...  ..
145   6.7  3.0    5.2   2.3   2
146   6.3  2.5    5.0   1.9   2
147   6.5  3.0    5.2   2.0   2
148   6.2  3.4    5.4   2.3   2
149   5.9  NaN    5.1   1.8   2

[150 rows x 5 columns]

In [ ]:
# 원핫인코딩 전에 타입을 category로 버꿔줌
new_iris['품종'] = new_iris['품종'].astype('category')
new_iris.dtypes

꽃잎길이      float64
꽃잎폭       float64
꽃받침길이     float64
꽃받침폭      float64
품종       category
dtype: object

In [ ]:
# 원핫인코딩
new_iris = pd.get_dummies(new_iris)
new_iris

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭  품종_0  품종_1  품종_2
0     5.1  3.5    1.4   0.2     1     0     0
1     4.9  3.0    1.4   0.2     1     0     0
2     4.7  3.2    1.3   0.2     1     0     0
3     4.6  3.1    1.5   0.2     1     0     0
4     5.0  3.6    1.4   0.2     1     0     0
..    ...  ...    ...   ...   ...   ...   ...
145   6.7  3.0    5.2   2.3     0     0     1
146   6.3  2.5    5.0   1.9     0     0     1
147   6.5  3.0    5.2   2.0     0     0     1
148   6.2  3.4    5.4   2.3     0     0     1
149   5.9  NaN    5.1   1.8     0     0     1

[150 rows x 7 columns]

In [ ]:
# na 값 체크
new_iris.isna().sum()

꽃잎길이     0
꽃잎폭      1
꽃받침길이    0
꽃받침폭     0
품종_0     0
품종_1     0
품종_2     0
dtype: int64

In [ ]:
# na 값에 꽃입폭 평균값 넣어주기
mean = new_iris['꽃잎폭'].mean()
new_iris['꽃잎폭'] = new_iris['꽃잎폭'].fillna(mean)
new_iris.tail()

꽃잎길이       꽃잎폭  꽃받침길이  꽃받침폭  품종_0  품종_1  품종_2
145   6.7  3.000000    5.2   2.3     0     0     1
146   6.3  2.500000    5.0   1.9     0     0     1
147   6.5  3.000000    5.2   2.0     0     0     1
148   6.2  3.400000    5.4   2.3     0     0     1
149   5.9  3.054362    5.1   1.8     0     0     1

# 모델 만들기 부록
## Batch Normalization
### 

## Boston housing price

In [ ]:
path = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/boston.csv'
boston = pd.read_csv(path)
boston.columns

Index(['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax',
       'ptratio', 'b', 'lstat', 'medv'],
      dtype='object')

In [ ]:
ind = boston[['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax',
       'ptratio', 'b', 'lstat']]
dep = boston[['medv']]

In [ ]:
# 모델 구조 만들기
X = tf.keras.layers.Input(shape=[13])
H = tf.keras.layers.Dense(8, activation='swish')(X)
H = tf.keras.layers.Dense(8, activation='swish')(H)
H = tf.keras.layers.Dense(8, activation='swish')(H)
Y = tf.keras.layers.Dense(1)(H)
model = tf.keras.models.Model(X,Y)
model.compile(loss='mse')

In [ ]:
# 모델 구조에 BatchNormilzation layer 추가
X = tf.keras.layers.Input(shape=[13])

H = tf.keras.layers.Dense(8)(X)
H = tf.keras.layers.BatchNormalization()(H)
H = tf.keras.layers.Activation('swish')(H)

H = tf.keras.layers.Dense(8)(H)
H = tf.keras.layers.BatchNormalization()(H)
H = tf.keras.layers.Activation('swish')(H)

H = tf.keras.layers.Dense(8)(H)
H = tf.keras.layers.BatchNormalization()(H)
H = tf.keras.layers.Activation('swish')(H)

Y = tf.keras.layers.Dense(1)(H)
model = tf.keras.models.Model(X,Y)
model.compile(loss='mse')

In [ ]:
model.fit(ind,dep,epochs=1000,verbose=0)
model.fit(ind,dep,epochs=100)

Epoch 1/100
16/16 [==============================] - 0s 2ms/step - loss: 9.9202
Epoch 2/100
16/16 [==============================] - 0s 3ms/step - loss: 13.1413
Epoch 3/100
16/16 [==============================] - 0s 2ms/step - loss: 9.0985
Epoch 4/100
16/16 [==============================] - 0s 3ms/step - loss: 9.6942
Epoch 5/100
16/16 [==============================] - 0s 3ms/step - loss: 8.2518
Epoch 6/100
16/16 [==============================] - 0s 3ms/step - loss: 9.2534
Epoch 7/100
16/16 [==============================] - 0s 3ms/step - loss: 7.7465
Epoch 8/100
16/16 [==============================] - 0s 2ms/step - loss: 8.7712
Epoch 9/100
16/16 [==============================] - 0s 2ms/step - loss: 9.5346
Epoch 10/100
16/16 [==============================] - 0s 2ms/step - loss: 9.5898
Epoch 11/100
16/16 [==============================] - 0s 2ms/step - loss: 7.7704
Epoch 12/100
16/16 [==============================] - 0s 2ms/step - loss: 8.0117
Epoch 13/100
16/16 [================

## iris

In [ ]:
path = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/iris.csv'
iris = pd.read_csv(path)
iris = pd.get_dummies(iris)
iris.columns

Index(['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭', '품종_setosa', '품종_versicolor',
       '품종_virginica'],
      dtype='object')

In [ ]:
ind = iris[['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭']]
dep = iris[['품종_setosa', '품종_versicolor',
       '품종_virginica']]


In [ ]:
# 모델 구조 생성(BatchNormalization 사용)
X = tf.keras.layers.Input(shape=[4])

H = tf.keras.layers.Dense(8)(X)
H = tf.keras.layers.BatchNormalization()(H)
H = tf.keras.layers.Activation('swish')(H)

H = tf.keras.layers.Dense(8)(X)
H = tf.keras.layers.BatchNormalization()(H)
H = tf.keras.layers.Activation('swish')(H)

H = tf.keras.layers.Dense(8)(X)
H = tf.keras.layers.BatchNormalization()(H)
H = tf.keras.layers.Activation('swish')(H)

Y = tf.keras.layers.Dense(3,activation='softmax')(H)

model = tf.keras.models.Model(X,Y)
model.compile(loss = 'categorical_crossentropy',metrics='accuracy')

model.fit(ind,dep,epochs=1000)

Epoch 1/1000
5/5 [==============================] - 1s 3ms/step - loss: 1.1010 - accuracy: 0.5200
Epoch 2/1000
5/5 [==============================] - 0s 3ms/step - loss: 1.0314 - accuracy: 0.6067
Epoch 3/1000
5/5 [==============================] - 0s 3ms/step - loss: 1.0159 - accuracy: 0.6000
Epoch 4/1000
5/5 [==============================] - 0s 2ms/step - loss: 0.9763 - accuracy: 0.6333
Epoch 5/1000
5/5 [==============================] - 0s 2ms/step - loss: 0.9418 - accuracy: 0.6400
Epoch 6/1000
5/5 [==============================] - 0s 6ms/step - loss: 0.9056 - accuracy: 0.6333
Epoch 7/1000
5/5 [==============================] - 0s 2ms/step - loss: 0.8915 - accuracy: 0.6400
Epoch 8/1000
5/5 [==============================] - 0s 1ms/step - loss: 0.8777 - accuracy: 0.6133
Epoch 9/1000
5/5 [==============================] - 0s 1ms/step - loss: 0.8414 - accuracy: 0.6267
Epoch 10/1000
5/5 [==============================] - 0s 2ms/step - loss: 0.8267 - accuracy: 0.6267
Epoch 11/1000
5/5 [

In [ ]:
list_ = [26.1,26.1,26,26.1,26.3,26.6,26.8,27.1,27.3,27.5]
len(list_)

10

In [ ]:
import statistics

statistics.mean(list_)

26.59

In [ ]:
statistics.pvariance(list_)

0.2788999999999998

In [ ]:
list__ = [21,21,21,27,45,50,49,44,39,36]
len(list__)

10

In [ ]:
statistics.pvariance(list__)

127.01

In [ ]:
statistics.pstdev(list__)

11.269871339105872

In [ ]:
list_tmp = [24.2,24.3,24.3,24.2,24.3,24.4,24.6,24.8,25.8,26.8]
statistics.mean(list_tmp)

24.77

In [ ]:
statistics.pvariance(list_tmp)

0.6661000000000005

In [ ]:
statistics.pstdev(list_tmp)

0.8161494961096285